### PR TITLE
Remove string mutations in Parser::Lexer::Literal.

### PR DESCRIPTION
**Pull request by iliabylich**
Required for https://github.com/opal/opal/pull/1465

### CODE REVIEW

**Comment by whitequark** [comment]
Well, no, absolutely not. Unlike in the case with the `ast` gem, where the replacement you did was in debug inspection code, this is one of the hottest paths. So if you want me to merge this, you are going to need a benchmark, at least.

File: lib/parser/lexer/literal.rb [hunk 1]
```diff
@@ -78,7 +78,7 @@ def initialize(lexer, str_type, delimiter, str_s, heredoc_e = nil,
                       !heredoc?)
 
       # Capture opening delimiter in percent-literals.
-      @str_type << delimiter if @str_type.start_with?('%'.freeze)
+      @str_type += delimiter if @str_type.start_with?('%'.freeze)
```
Comment (whitequark) [inline lib/parser/lexer/literal.rb]: To be more precise: I have no objections to merging this chunk, only the other one.

**Comment by iliabylich** [comment]
@whitequark Can I use `bin/ruby-parse -B file.rb`?

**Comment by whitequark** [comment]
On a good enough corpus. At least the Ruby stdlib, and also some Ruby gem with a massive amount of strings. (I don't remember what it was, sorry. Please search the bug tracker, and tell me if you can't find the issue that mentions it.)

**Comment by iliabylich** [comment]
@whitequark Alright, I've used [all these files](https://github.com/ruby/spec/tree/master/core/string) for testing. (I can bm the whole rubyspec test suite if you want)
Master:
```
rake regenerate
bin/ruby-parse --23 -B ../opal/spec/ruby/core/string/
# => Parsed 125 files (373273 characters) in 2.22 seconds (168.141 kchars/s).
# => Parsed 125 files (373273 characters) in 2.21 seconds (168.902 kchars/s).
```
Patch:
```
rake regenerate
bin/ruby-parse --23 -B ../opal/spec/ruby/core/string/
# => Parsed 125 files (373273 characters) in 2.23 seconds (167.387 kchars/s).
# => Parsed 125 files (373273 characters) in 2.22 seconds (168.141 kchars/s).
```

**Comment by whitequark** [comment]
No, that doesn't really stress `extend_string`, since what really matters (that I neglected to mention...) is the ratio of string characters to other tokens.

**Comment by whitequark** [comment]
Okay, I cannot find the exact gem, but it's easy enough to reproduce. Please generate a file with about 1MB of Ruby strings and nothing else, with maybe 20% of the content in those strings being random Unicode escape sequences. That will be very representative.

**Comment by iliabylich** [comment]
(Just in case someone may need it in future)
``` ruby
f = File.open((Dir.pwd + "/test.rb"), 'w')
8000.times do
  f.write( ("a" * 100).inspect )
  f.write("\n")
end
2000.times do
  f.write( '"' + '\u0123' * 100 + '"' )
  f.write("\n")
end
f.close
```

```
wc test.rb
  10000   10000 2030000 test.rb
```

Master:
```
bin/ruby-parse -B test.rb
# => Parsed 1 files (2030000 characters) in 4.07 seconds (498.771 kchars/s).
# => Parsed 1 files (2030000 characters) in 4.08 seconds (497.549 kchars/s).
# => Parsed 1 files (2030000 characters) in 4.12 seconds (492.718 kchars/s).
# => Parsed 1 files (2030000 characters) in 4.10 seconds (495.122 kchars/s).
```

Patch
```
bin/ruby-parse -B test.rb
# => Parsed 1 files (2030000 characters) in 4.16 seconds (487.981 kchars/s).
# => Parsed 1 files (2030000 characters) in 4.30 seconds (472.093 kchars/s).
# => Parsed 1 files (2030000 characters) in 4.16 seconds (487.981 kchars/s).
# => Parsed 1 files (2030000 characters) in 4.28 seconds (474.299 kchars/s).
```

**Comment by whitequark** [comment]
I guess that's fine.

**Pull request merged**